### PR TITLE
Users: Reapply hostfilter locks after user merge

### PR DESCRIPTION
**Pull request by mia-pi-git**

### CODE REVIEW

**Comment by asgdf** [comment]
It should probably only check latestIp, otherwise we would go back to people having to log out and refresh after getting off proxies

**Comment by Zarel** [comment]
@mia-pi-git?

**Comment by mia-pi-git** [comment]
Oh, we talked about this and then i got wrapped up in other things. I'll commit that now, it's pretty easy.

File: server/users.ts [hunk 1]
```diff
@@ -997,6 +997,9 @@ export class User extends Chat.MessageContext {
 		this.userMessage = oldUser.userMessage || this.userMessage || '';
 
 		oldUser.markDisconnected();
+
+		const latestConn = this.connections.find(c => c.ip === this.latestIp);
+		void Punishments.checkIp(this, latestConn!);
```
Comment (Zarel) [inline server/users.ts:1002]:
```suggestion
		const latestConnection = this.connections[this.connections.length - 1];
		if (latestConnection) void Punishments.checkIp(this, latestConnection);
```

**Comment by Zarel** [comment]
Wait, why is this necessary at all? Don't the locks already apply to the individual connections?

**Comment by Zarel** [comment]
After some discussion in Staff, I think this is the wrong direction:

[14:51] #Zarel: like, you should only be hostfilter locked if all your connections are hostfilter locked
[14:51] #Zarel: we can mark the connections with hostfilter lock status if that would make it easier

**Comment by asgdf** [comment]
I'd just like to point out that I consider this a very high-priority fix, on account of the entire hostfilter system not working at all at the moment

**Comment by Zarel** [comment]
> I'd just like to point out that I consider this a very high-priority fix, on account of the entire hostfilter system not working at all at the moment

That's awkward because this requires a restart to apply which is definitely not happening today...

**Comment by Zarel** [comment]
Actually, it might not.......

**Comment by mia-pi-git** [comment]
> > I'd just like to point out that I consider this a very high-priority fix, on account of the entire hostfilter system not working at all at the moment
> 
> That's awkward because this requires a restart to apply which is definitely not happening today...

Well, this can be monkeypatched / hotpatched (ish?) - just would be difficult.

**Comment by asgdf** [comment]
To be clear we'll survive some more days if need be, I just wanted to caution against the regression bugfix turning into a refactor PR

**Comment by asgdf** [comment]
It seems Annika has pushed a fix for now (eceb8549ddfc26ddaf1a9c6810455357296d6626), so this is no longer as urgent - refactor away!

**Comment by Zarel** [comment]
OK. I think this should be closed since it is bad code. The current code is pretty fucked up but this just piles more hacks upon hacks instead of being a proper refactor.

**Comment by Zarel** [comment]
In particular, I'm unhappy with this `if` statement:

```ts
// If either user is unlocked and neither is locked by name, remove the lock.
// Otherwise, keep any locks that were by name.
if (
	(!oldUser.locked || !this.locked) &&
	oldUser.locked !== oldUser.id &&
	this.locked !== this.id &&
	// Only unlock if no previous names are locked
	!oldUser.previousIDs.some(id => !!Punishments.search(id)
		.filter(punishment => punishment[2][0] === 'LOCK' && punishment[2][1] === id)
		.length)
) {
```

**Comment by Zarel** [comment]
We should not need to do five Punishment searches to merge two users. :[

**Comment by Zarel** [comment]
I think the solution is for the user to track if it's been "strongly locked" or "weakly locked".

**Pull request closed without merging**